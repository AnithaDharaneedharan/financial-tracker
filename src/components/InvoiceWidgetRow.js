import React, { useState, useEffect } from "react";
import { Button } from "react-bootstrap";
import { dateFormatter } from "../utils/dateFormatter";
//import { getData } from "../services/localData";

export default function InvoiceWidgetRow({ invoice, onRemove, onSaveEvent, transactionDetails }) {
  const [referenceNumber, setReferenceNumber] = useState("");
  const [customerName, setCustomerName] = useState("");
  const [paymentStatus, setPaymentStatus] = useState("NOT PAID");
  const [creationDate, setCreationDate] = useState(dateFormatter(new Date()));
  const [amount, setAmount] = useState(0);

  // const localStorageData = getData("data");
  // const { transactionDetails } = localStorageData;

  useEffect(() => {
    setReferenceNumber(invoice.referenceNumber);
    setCustomerName(invoice.customerName);
    setCreationDate(invoice.creationDate);
    setAmount(invoice.amount);
  }, [invoice]);

  useEffect(() => {
    const _paymentStatus = transactionDetails.find(
      (transactionDetail) =>
        invoice.referenceNumber === transactionDetail.referenceNumber &&
        invoice.amount === transactionDetail.amount &&
        new Date(transactionDetail.date) >= new Date(invoice.creationDate)
    )
      ? "PAID"
      : "NOT PAID";
    setPaymentStatus(_paymentStatus);
  }, [invoice, transactionDetails]);


  return (
    <tr>
      <td>
        <input
          type="text"
          value={referenceNumber || ""}
          disabled={paymentStatus === "PAID"}
          onChange={(e) => setReferenceNumber(e.target.value)}
        ></input>
      </td>
      <td>
        <input
          type="text"
          value={customerName}
          disabled={paymentStatus === "PAID"}
          onChange={(e) => setCustomerName(e.target.value)}
        ></input>
      </td>
      <td>
        <input
          type="text"
          value={creationDate}
          disabled={paymentStatus === "PAID"}
          onChange={(e) => setCreationDate(e.target.value)}
        ></input>
      </td>

      <td>
        <input
          type="text"
          value={amount}
          disabled={paymentStatus === "PAID"}
          onChange={(e) => setAmount(e.target.value)}
        ></input>
      </td>

      <td>
        <span className={paymentStatus === "PAID" ? "paid" : "unpaid"}>
          {paymentStatus}
        </span>
      </td>
      <td>
        {paymentStatus !== "PAID" ? (
          <span className="action">
            <Button variant="secondary" onClick={() => onSaveEvent({referenceNumber,customerName, creationDate, amount, paymentStatus})}>
              Save
            </Button>
            <Button variant="secondary" onClick={() => onRemove(invoice)}>
              Remove
            </Button>
          </span>
        ) : (
          <span>N/A</span>
        )}
      </td>
    </tr>
  );
}
