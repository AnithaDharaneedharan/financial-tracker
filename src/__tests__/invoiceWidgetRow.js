import "@testing-library/jest-dom";
import React from "react";
import {
  render,
  fireEvent,
  cleanup,
} from "@testing-library/react";
import InvoiceWidgetRow from "../components/InvoiceWidgetRow";
import mockbankAccDetails from "../services/bankAccDetails.json";

const mockedInvoice = {
  referenceNumber: "UBER123",
  customerName: "UBER",
  billingAddress: "SM6 1RD, LONDON",
  creationDate: "2020-08-08",
  dueDate: "2020-09-08",
  amount: -40,
  customerAccountNumber: "34476ERT5",
  customerSortCode: "9876",
};

afterEach(cleanup);

test("should render the component", () => {
  const { asFragment } = render(
    <InvoiceWidgetRow
      transactionDetails={mockbankAccDetails.transactionDetails}
      invoice={mockedInvoice}
      onSaveEvent={() => {}}
      onRemove={() => {}}
    />
  );
  expect(asFragment()).toMatchSnapshot();
});

test.skip('it should render InvoiceWidgetRow reference number', () => {
    const { getByTestId } = render(<InvoiceWidgetRow invoice={mockedInvoice}/>);
    const input = getByTestId('reference-number');
    fireEvent.change(input, { target : {value: 'ANI345'}});
    expect(input.value).toBe('ANI345');
  })
  
