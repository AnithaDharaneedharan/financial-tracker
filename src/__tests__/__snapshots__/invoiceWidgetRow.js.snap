// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the component 1`] = `
<DocumentFragment>
  <tr>
    <td>
      <input
        data-testid="reference-number"
        type="text"
        value="UBER123"
      />
    </td>
    <td>
      <input
        data-testid="customer-name"
        type="text"
        value="UBER"
      />
    </td>
    <td>
      <input
        data-testid="creation-date"
        type="text"
        value="2020-08-08"
      />
    </td>
    <td>
      <input
        data-testid="amount"
        type="number"
        value="-40"
      />
    </td>
    <td>
      <span
        class="unpaid"
      >
        NOT PAID
      </span>
    </td>
    <td>
      <span
        class="action"
      >
        <button
          class="btn btn-secondary"
          type="button"
        >
          Save
        </button>
        <button
          class="btn btn-secondary"
          type="button"
        >
          Remove
        </button>
      </span>
    </td>
  </tr>
</DocumentFragment>
`;
